Implement name inference for mm, addmm, addmm_

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25181 [WIP][testing] Turn on BUILD_NAMEDTENSOR=1
* #25178 Include the correct header for make_unique in named tensor headers
* #25177 Implement name inference for torch.matmul
* #25123 Implement name inference for torch.bmm
* #24474 Implement name inference for torch.dot
* #24471 Implement name inference for addmv, addmv_, mv
* #24469 Implement name inference for expand
* **#24306 Implement name inference for mm, addmm, addmm_**
* #24921 Fix binary op name inference between unnamed and named tensors.

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

Differential Revision: [D16915801](https://our.internmc.facebook.com/intern/diff/D16915801)

zou3519 committed:
Implement name inference for mm, addmm

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.
- The out= variants aren't implemented. This is because they are a
little weird: the output gets resized, but I haven't throught through
what semantics that should have.

Test Plan:
- new tests [namedtensor ci]

zou3519 committed:
Update on "Implement name inference for mm, addmm"

Implement name inference for mm, addmm

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.
- The out= variants aren't implemented. This is because they are a
little weird: the output gets resized, but I haven't throught through
what semantics that should have.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

nairbv commented on test/test_namedtensor.py:
```diff
@@ -23,10 +24,58 @@ def flatten(lst):
+        casted_args = [arg.to(device) if isinstance(arg, torch.Tensor) else arg
+                       for arg in args]
+        if maybe_raises_regex is not None:
+            with self.assertRaisesRegex(RuntimeError, maybe_raises_regex):
```
i don't see this used... should we test some cases that fail (e.g. mis-matched names in a mm op should fail right?)

zou3519 replied:
Right now, the name inference rule for mm doesn't fail if names are mismatched, so that won't fail.

nairbv replied:
is it supposed to fail?

zou3519 replied:
The decision right now is for it to not fail. Tensor[M, N] @ Tensor[O, P] returns Tensor[M, P] and it doesn't matter that N / O are "mismatched".

nairbv replied:
ah now I understand, thanks

zou3519 committed:
Update on "Implement name inference for mm, addmm"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm, addmm_"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm, addmm_"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm, addmm_"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm, addmm_"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm, addmm_"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm, addmm_"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm, addmm_"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm, addmm_"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm, addmm_"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

zou3519 committed:
Update on "Implement name inference for mm, addmm, addmm_"

Implement name inference for mm, addmm, addmm_

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head
